File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_9.java
package org.twyszomirski.aoc.solutions;


import org.twyszomirski.aoc.utils.Tuple;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.io.IO.println;

public class Day_9 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_9.txt"));
        println("======== Day 9 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = readData(lines);
        var asLinked = new LinkedList<>(data);
        var nextEmptyIdx = 1;
        var nextFileIdx = asLinked.size() - 1;

        while (nextEmptyIdx < nextFileIdx) {
            var file = nextFile(asLinked, nextFileIdx);
            var empty = nextEmpty(asLinked, nextEmptyIdx);
            nextFileIdx = file.right();
            nextEmptyIdx = empty.right();
            empty.left().size--;
            file.left().size--;

            var prevFile = asLinked.get(empty.right()-1);
            if (prevFile.fileId.equals(file.left().fileId)) {
                prevFile.size++;
            } else {
                asLinked.add(nextEmptyIdx,new DiskBlock(1, file.left().fileId));
                nextFileIdx++;
            }
        }

        var onlyFiles = asLinked.stream().filter(it -> !it.isEmpty())
                .flatMap(file -> IntStream.range(0, file.size).map(a -> file.fileId).boxed())
                .toList();
        var sum = IntStream.range(0, onlyFiles.size()).boxed().mapToLong(idx -> (long) idx * onlyFiles.get(idx)).sum();
        println(sum);
    }

    private static void part_2(List<String> lines) {
        var data = readData(lines);
        var asLinked = new LinkedList<>(data);

        var nextFileIdx = asLinked.size() - 1;
        var currFileId = asLinked.size();
        while (nextFileIdx > 0) {
            var file = nextFile(asLinked, nextFileIdx, currFileId);
            currFileId = file.left().fileId;
            nextFileIdx = file.right();
            var empty = matchingEmpty(asLinked, file.left().size, file.right());
            if(empty != null){
                if(empty.left().size == file.left().size) {
                    empty.left().fileId = file.left().fileId;
                }else{
                    empty.left().size-=file.left().size;
                    asLinked.add(empty.right(), new DiskBlock(file.left().size, currFileId));
                }
                file.left().fileId = null;
            }
        }

        var onlyFiles = asLinked.stream()
                .flatMap(file -> IntStream.range(0, file.size).map(a -> file.fileId !=null ? file.fileId : 0 ).boxed())
                .toList();
        var sum = IntStream.range(0, onlyFiles.size()).boxed().mapToLong(idx -> (long) idx * onlyFiles.get(idx)).sum();
        println(sum);
    }

    private static void print(List<DiskBlock> data) {
        println(data.stream().map(DiskBlock::toString).collect(Collectors.joining("")));
    }

    private static Tuple<DiskBlock, Integer> nextEmpty(List<DiskBlock> data, Integer fromIdx) {
        for (int i = fromIdx; i < data.size(); i++) {
            var next = data.get(i);
            if (next.isEmpty() && next.size > 0) {
                return new Tuple<>(next, i);
            }
        }
        return null;
    }

    private static Tuple<DiskBlock, Integer> matchingEmpty(List<DiskBlock> data, Integer size, Integer maxIndex) {
        for (int i = 0; i < maxIndex; i++) {
            var next = data.get(i);
            if (next.isEmpty() && next.size >= size) {
                return new Tuple<>(next, i);
            }
        }
        return null;
    }

    private static Tuple<DiskBlock, Integer> nextFile(List<DiskBlock> data, Integer fromIdx) {
        for (int i = fromIdx; i > 0; i--) {
            var next = data.get(i);
            if (!next.isEmpty() && next.size > 0) {
                return new Tuple<>(next, i);
            }
        }
        return null;
    }

    private static Tuple<DiskBlock, Integer> nextFile(List<DiskBlock> data, Integer fromIdx, Integer maxFileId) {
        for (int i = fromIdx; i >= 0; i--) {
            var next = data.get(i);
            if (!next.isEmpty() && next.size > 0 && next.fileId < maxFileId) {
                return new Tuple<>(next, i);
            }
        }
        return null;
    }

    private static List<DiskBlock> readData(List<String> lines) {
        var digits = Arrays.stream(lines.getFirst().split(""))
                .mapToInt(Integer::valueOf).boxed().toList();
        return IntStream.range(0, digits.size()).mapToObj(idx ->
                (idx % 2 == 0) ? new DiskBlock(digits.get(idx), idx / 2) : new DiskBlock(digits.get(idx), null)
        ).toList();
    }

    static class DiskBlock {
        Integer size;
        Integer fileId;

        public DiskBlock(Integer size, Integer fileId) {
            this.size = size;
            this.fileId = fileId;
        }

        Boolean isEmpty() {
            return fileId == null;
        }

        @Override
        public String toString() {
            var symbol = isEmpty() ? "." : fileId.toString();
            return size == 0 ? "_" : IntStream.range(0, size).mapToObj(it -> symbol).collect(Collectors.joining());
        }

    }
}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_9.kt
package org.twyszomirski.aoc.solutions

import java.io.File
import java.util.LinkedList


class Day_9 {

    fun solve() {
        val lines = File("src/main/resources/input_day_9.txt").readLines()
        println("======== Day 9 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val diskData = getData(input)

        var currentEmptyPointer = 1
        var currentFilePointer = diskData.size - 1

        while (currentEmptyPointer <= currentFilePointer) {
            val (file, pointer) = nextFile(diskData, currentFilePointer)
            currentEmptyPointer = nextEmpty(diskData, currentEmptyPointer)
            currentFilePointer = pointer

            val prevFile = diskData[currentEmptyPointer - 1]
            if (prevFile.fileId == file.fileId) {
                prevFile.length++
            } else {
                diskData.add(currentEmptyPointer, DiskPart(1, file.fileId))
                currentEmptyPointer++
                currentFilePointer++
            }
            diskData[currentEmptyPointer].length--

            if (diskData[currentEmptyPointer].length == 0) {
                diskData.removeAt(currentEmptyPointer)
                currentFilePointer--
            }
        }

        val sum = diskData.filter { it.fileId != null }
            .flatMap { file -> (0..<file.length).map { file.fileId!! } }
            .foldIndexed(0L) { idx, acc, id ->
                acc + (idx * id)
            }

        println(sum)
    }

    fun part2(input: List<String>) {
        val data = getData(input)

        var currentFilePointer = data.size - 1

        while (currentFilePointer > 0) {
            val file = data[currentFilePointer]
            val currentFileId = file.fileId
            val emptyIdx = matchingEmpty(data, file.length, currentFilePointer)
            if (emptyIdx > 0) {
                val empty = data[emptyIdx]
                if (empty.length == file.length) {
                    data.removeAt(emptyIdx)
                    data.add(emptyIdx, DiskPart(file.length, file.fileId))
                } else {
                    data.add(emptyIdx, DiskPart(file.length, file.fileId))
                    empty.length -= file.length
                    currentFilePointer++
                }
                data[currentFilePointer].fileId = null
            }

            while (currentFilePointer > 0) {
                currentFilePointer--
                if (data[currentFilePointer].fileId != null && data[currentFilePointer].fileId!! < currentFileId!!) {
                    break
                }
            }
        }

        val sum = data.flatMap { file -> (0..<file.length).map { file.fileId ?: 0 } }
            .foldIndexed(0L) { idx, acc, id ->
                acc + (idx * id)
            }
        println(sum)

    }

    fun getData(lines: List<String>): LinkedList<DiskPart> {
        return lines.first().split("").filter {
            it.isNotBlank()
        }.map { it.toInt() }.mapIndexed { idx, size ->
            val fileId = if (idx % 2 == 0) idx / 2 else null
            DiskPart(size, fileId)
        }.toCollection(LinkedList())
    }


    fun nextFile(diskData: MutableList<DiskPart>, currPointer: Int): Pair<DiskPart, Int> {
        val file = diskData[currPointer]
        if (file.length == 1) {
            diskData.removeAt(currPointer)
            return file to currPointer - 2
        } else {
            file.length--
            return file to currPointer
        }
    }

    fun nextEmpty(diskData: MutableList<DiskPart>, currPointer: Int): Int {
        return (currPointer - 1) + diskData.drop(currPointer - 1).indexOfFirst { it.fileId == null && it.length > 0 }
    }

    fun matchingEmpty(data: List<DiskPart>, size: Int, maxId: Int): Int {
        return data.subList(0, maxId).indexOfFirst { it.fileId == null && it.length >= size }
    }

    data class DiskPart(var length: Int, var fileId: Int?) {
        override fun toString(): String {
            val symbol = fileId?.toString() ?: "."
            return (0..<length).joinToString("") { symbol }
        }
    }


}